libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/DownloadTask.kt

```diff
@@ -0,0 +1,100 @@
+package com.mapbox.services.android.navigation.v5.utils
+
+import android.os.AsyncTask
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import java.io.InputStream
+import java.io.OutputStream
+import okhttp3.ResponseBody
+
+/**
+ * This class is an [AsyncTask] that downloads a file from a [ResponseBody].
+ * Creates a DownloadTask which will download the input stream into the given
+ * destinationDirectory with the specified file name and file extension. If a listener is passed
+ *
+ * @param destDirectory path to the directory where file should be downloaded
+ * @param fileName name to name the file
+ * @param extension file extension of the resulting file
+ * @param downloadListener listener to be updated on completion of the task
+ */
+class DownloadTask
+@JvmOverloads
+constructor(
+    private val destDirectory: String,
+    private val fileName: String = "",
+    private val extension: String,
+    private val downloadListener: DownloadListener
+) : AsyncTask<ResponseBody, Void, File>() {
+
+    companion object {
+        private const val END_OF_FILE_DENOTER = -1
+        private var uniqueId = 0
+    }
+
+    override fun doInBackground(vararg responseBodies: ResponseBody): File? =
+        saveAsFile(responseBodies.firstOrNull())
+
+    /**
+     * Saves the file returned in the response body as a file in the cache directory
+     *
+     * @param responseBody containing file
+     * @return resulting file, or null if there were any IO exceptions
+     */
+    private fun saveAsFile(responseBody: ResponseBody?): File? {
+        if (responseBody == null) {
+            return null
+        }
+
+        try {
+            val file =
+                File(destDirectory + File.separator + fileName + retrieveUniqueId() + "." + extension)
```

Comment:
_String interpolation_?

Reply:
done

Reply:
I was thinking something around these lines

`File("$destDirectory${File.separator}$fileName${retrieveUniqueId()}.$extension")`

`StringBuilder` also works, we can fix later in a later PR 👍